### PR TITLE
Provide parsing for ServiceEndpoint

### PR DESCRIPTION
**Pull request by jskeet**
(Assigned to @chrisdunelm rather than @amanda-tarafa somewhat randomly...)

### CODE REVIEW

File: Google.Api.Gax.Grpc.Tests/ServiceEndpointTest.cs [hunk 1]
```diff
@@ -84,5 +84,50 @@ public void Equality()
...
+        [InlineData("   host:443")]
+        [InlineData("host:443    ")]
+        [InlineData("host: 443")]
+        [InlineData("host    ")]
+        [InlineData("    host")]
+        public void Parse_Invalid(string text)
+        {
+            Assert.False(ServiceEndpoint.TryParse(text, out var endpoint));
+            Assert.Null(endpoint);
+            Assert.Throws<ArgumentException>(() => ServiceEndpoint.Parse(text));
```
Comment (chrisdunelm) [inline Google.Api.Gax.Grpc.Tests/ServiceEndpointTest.cs]: super-nit: These asserts are not in the same order as in `Parse_Null()`!

File: Google.Api.Gax.Grpc.Tests/ServiceEndpointTest.cs [hunk 2]
```diff
@@ -84,5 +84,50 @@ public void Equality()
...
+        public void Parse_Valid(string text, string expectedHost, int expectedPort)
+        {
+            Assert.True(ServiceEndpoint.TryParse(text, out var endpoint));
+            Assert.Equal(expectedHost, endpoint.Host);
+            Assert.Equal(expectedPort, endpoint.Port);
+
+            Assert.Equal(endpoint, ServiceEndpoint.Parse(text));
+        }
+
+        [Theory]
```
Comment (chrisdunelm) [inline Google.Api.Gax.Grpc.Tests/ServiceEndpointTest.cs]: I'd suggest adding the testcase of ":"

Comment (jskeet) [reply Google.Api.Gax.Grpc.Tests/ServiceEndpointTest.cs]: Done.

File: Google.Api.Gax.Grpc.Tests/ServiceEndpointTest.cs (hunk 1 above)
Comment (jskeet) [reply Google.Api.Gax.Grpc.Tests/ServiceEndpointTest.cs]: Done.

**Pull request merged**